### DISCUSSION TITLE
We need to use a better structure for results of simulation

### DISCUSSION THREAD

**Discussion by ElCep** [cormas/cormas#773] [Ideas]
Working on the OpenMole interaction, I need to extract data from models in JSON.
Thanks to @pierbo I have discovered `data` object but its structure is quite strange for me. With data look like 
```
a Dictionary('CMFireAutomataModel'->a Dictionary('nbAshes'->193 'nbFires'->851 ) )
"and not"
a Dictionary('nbAshes'->193 'nbFires'->851)
```
What do you think about using data frame (From [DataFrame](https://github.com/PolyMathOrg/DataFrame#1-creating-dataframe-from-an-array-of-rows-or-columns)) for data manipulation ?
E.

**Comment by nbecu** [comment]
This structure is because Cormas allows to have probes on the attributes or
states of model entities.

For instance in ECEC you can save the energy of the Forager class.
In this case, the energy of the instances of the Foragers will be saved. The data will be stored in the data object as so
 Dictionary(
 'CMECECModel'->a Dictionary('nbFiremen'->5 ) 
'CMECECForager'->a Dictionary('energy'->12)
)


Le 31 mai 2018 6:22 PM, "Etienne DELAY" <notifications@github.com> a écrit :

Working on the OpenMole interaction, I need to extract data from models in
JSON.
Thanks to @pierbo <https://github.com/pierbo> I have discovered data object
but its structure is quite strange for me. With data look like

a Dictionary('CMFireAutomataModel'->a Dictionary('nbAshes'->193
'nbFires'->851 ) )
"and not"
a Dictionary('nbAshes'->193 'nbFires'->851)

What do you think about using data frame (From DataFrame
<https://github.com/PolyMathOrg/DataFrame#1-creating-dataframe-from-an-array-of-rows-or-columns>)
for data manipulation ?
E.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/cormas/cormas/issues/56>, or mute the thread

**Comment by SergeStinckwich** [comment]
@nbecu your reply is not complete.

**Comment by SergeStinckwich** [comment]
This question happens because when @ElCep try to use the result of a simulation and want to access only to the last run of a simulation. The resulting code was very tricky.

Using a dictionnary of dictionnary of collection for storing resulting data of a simultion is not a good OO design, because everytime you need to do an operation on this structure, this is tricky, because you need to know how the dictionnary is structured. We need to use real objects here.

Using DataFrame will allow us to do complex operations very easily like filtering, selecting rows or columns.

**Comment by pierbo** [comment]
completely agree !  :)

**Comment by nbecu** [comment]
Sure . Let's do that

Le jeu. 31 mai 2018 22:45, Serge Stinckwich <notifications@github.com> a
écrit :

**Comment by SergeStinckwich** [comment]
I try to understand the structure of the actual Dictionnary. You have global data that are stored with the model class key and information about each classes ? right ?

**Comment by nbecu** [comment]
yes.
global data that are stored with the model class key
local probes data are stored  with the associated entity class key

2018-06-01 12:53 GMT+02:00 Serge Stinckwich <notifications@github.com>:

> I try to understand the structure of the actual Dictionnary. You have
> global data that are stored with the model class key and information about
> each classes ? right ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cormas/cormas/issues/56#issuecomment-393845758>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHRtZyzKP-JfQg0gXeDxzkBAwUlx5Vptks5t4R0bgaJpZM4UVWUS>
> .
>



-- 
__________________
Nicolas Becu
Chercheur au CNRS
Littoral Environnement et Sociétés (LIENSs) - UMR 7266
Bâtiment ILE
2, rue Olympe de Gouges
17 000 La Rochelle
Tel : 05 16 49 67 94

**Comment by nbecu** [comment]
This is how it looks like in VW version
in this ex. we have 2 global probes stored ( restrainedSize  and
vegetationBiomass )
And 1 local probe stored (the energy of the Restrained agent entity)
                  / note that there are 10 restrained agent instances

In this example the simulation was run for 12 steps

Dictionary (
#ECEC->Dictionary (#restrainedSize->List (10 10 10 10 10 10 10 10 10 10 10
10 10)
                                #vegetationBiomass->List (377.789 440.911
510.405 589.681 677.379 767.917 870.332 981.46 1102.32 1233.95 1373.32
1520.65 1674.31)
                                )
 #Restrained->Dictionary (#energy->Dictionary (
                                     1->OrderedCollection (45 43.6851
42.4677 41.2079 39.8559 38.9927 38.1138 36.4023 35.8137 35.4269 35.1769
34.8916 34.7312)
                                     2->OrderedCollection (45 43.2124
41.5058 40.2658 38.5262 37.7022 36.1698 35.4337 35.0098 34.8289 34.5248
34.672 34.9073)
                                    3->OrderedCollection (45 43.5704
42.1708 40.6308 39.5561 38.6673 37.8124 37.0953 36.7155 36.5421 36.0891
35.9178 36.3249)
                                   4->OrderedCollection (45 43.2775 41.9209
40.3108 38.8039 37.9261 37.1386 36.5667 35.8466 35.1261 34.8194 34.7769
34.613)
                                     5->OrderedCollection (45 43.4307
41.7358 40.3057 39.3169 38.4369 37.6485 36.9528 36.5805 36.3011 35.9363
35.6996 35.6237)
                                     6->OrderedCollection (45 43.2262
41.4271 40.2103 39.1076 38.257 36.9944 36.3847 35.0641 34.5636 33.9294
33.0934 32.5598)
                                   7->OrderedCollection (45 43.6572 41.9719
40.7301 39.4589 38.54 37.7499 37.0077 36.1815 35.6729 35.1885 34.9625
35.3337)
                                     8->OrderedCollection (45 43.6591
42.3084 40.8637 39.6504 38.7153 37.8803 37.3225 36.9322 36.6938 36.737
36.8394 36.989)
                                    9->OrderedCollection (45 43.554 42.0229
40.4983 39.2777 38.3513 37.4055 36.8451 36.4481 35.8815 35.2461 34.635
34.5772)
                                   10->OrderedCollection (45 43.2175
41.8658 40.6481 39.2315 38.406 37.5888 36.8058 36.4656 36.2002 36.1146
36.2872 36.5025)
                                    ) )
 )





Here is the same example with 90 steps of simulation (some agents of the
begining have disapear and some new agents have arrived in the course of
the simulation)

Dictionary (
#ECEC->Dictionary (#restrainedSize->List (10 10 10 10 10 10 10 10 10 10 10
10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10 10
10 10 10 10 10 10 10 10 11 11 13 13 13 13 13 13 13 14 14 14 14 14 14 14 14
14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 14 13 13 13
13 12 12 8 8) #vegetationBiomass->List (377.789 440.911 510.405 589.681
677.379 767.917 870.332 981.46 1102.32 1233.95 1373.32 1520.65 1674.31
1833.7 1994.3 2157.6 2323.22 2491.04 2662.38 2825.18 2980.07 3130.68
3279.45 3423.9 3562.6 3664.28 3748.92 3822.24 3875.61 3913.69 3946.44
3972.43 3978.24 3963.12 3921.58 3848.87 3766.64 3677.26 3596.77 3488.3
3367.15 3235.47 3057.66 2817.95 2618.65 2427.09 2216.1 2033.63 1846.5
1685.3 1533.17 1365.56 1198.86 1072.72 937.167 821.119 711.721 643.025
583.673 533.706 494.747 453.729 416.717 381.21 361.699 333.315 321.397
301.955 285.943 270.803 250.865 238.282 238.716 225.926 208.299 202.537
197.232 187.756 183.164 179.241 163.566 155.567 150.914 148.232 144.045
145.084 140.057 143.569 142.885 142.403 139.769) )
 #Restrained->Dictionary (#energy->Dictionary (1->OrderedCollection (45
43.6851 42.4677 41.2079 39.8559 38.9927 38.1138 36.4023 35.8137 35.4269
35.1769 34.8916 34.7312 34.9504 35.5592 36.1843 37.0463 38.0684 38.8817
39.747 41.0895 42.6149 43.877 45.0702 46.0693 47.0131 47.3839 47.5884
48.8467 49.9288 51.6526 53.6324 55.6647 57.7607 59.8663 61.9885 63.9759
66.1123 68.1866 69.8947 71.8885 73.7027 75.7904 77.882 79.9121 81.8441
83.8759 85.8906 87.8995 88.6995 89.3902 89.5015 89.8522 45.2574 45.0128
44.9304 44.2695 43.6426 42.8769 42.0628 41.1721 40.275 39.3391 37.8183
36.6736 35.3074 33.5231 31.8349 30.0473 28.2911 26.73 24.9417 23.8978
23.3003 22.4433 21.5398 20.1275 19.2413 17.959 16.932 15.2516 13.9489
12.1489 10.349 8.54899 6.75121 4.98145 3.47692 1.67696)
2->OrderedCollection (45 43.2124 41.5058 40.2658 38.5262 37.7022 36.1698
35.4337 35.0098 34.8289 34.5248 34.672 34.9073 35.2759 35.9427 36.6191
37.3799 38.4162 39.3156 40.7595 42.3008 43.5862 45.0376 46.4447 48.3891
50.4019 52.3223 54.1925 56.2632 58.3216 60.3524 62.4032 64.4503 66.5484
68.6535 70.7275 72.8424 74.7385 76.5003 78.3866 80.4761 82.5067 84.1963
86.0141 87.9967 87.7959 88.4727 88.4642 88.3959 88.3713 89.1513 89.0827
88.5188 87.9872 87.7676 87.2065 86.5624 85.6842 84.7479 83.7756 82.8678
81.9092 80.9719 80.0256 79.1401 78.2076 77.5392 75.8756 75.1663 73.4221
72.4398 71.0416 69.3178 67.5224 66.664 64.8988 63.1585 61.366 59.6132
57.9317 56.996 55.2893 53.4906 52.5188 51.7996 50.9806 50.5909 49.8614
49.1017 47.5021 46.2012) 3->OrderedCollection (45 43.5704 42.1708 40.6308
39.5561 38.6673 37.8124 37.0953 36.7155 36.5421 36.0891 35.9178 36.3249
36.8606 37.6524 38.5201 39.6253 40.7078 41.1603 42.431 43.2968 45.0196
46.4651 47.8884 49.5052 51.208 53.1962 55.2429 57.3354 59.3365 61.37
62.6361 64.1271 65.784 67.6205 69.5485 71.0149 72.7656 73.1605 73.7126
74.2731 75.0375 75.766 76.0413 76.4896 77.8649 79.7187 81.6852 83.3208
85.2754 86.1145 85.8513 84.9851 84.1149 83.2456 83.2352 82.1386 81.6338
80.3715 78.719 76.9679 76.0106 75.2594 74.3007 73.2479 72.3624 71.2967
69.6252 68.6313 67.7463 66.5594 64.9138 63.3351 62.4705 60.6977 59.6045
58.6855 56.9113 55.1614 53.7525 52.8567 51.0919 49.9762 48.3361 46.5418
44.7435 42.948 41.1485 39.3578 37.56 35.7603) 4->OrderedCollection (45
43.2775 41.9209 40.3108 38.8039 37.9261 37.1386 36.5667 35.8466 35.1261
34.8194 34.7769 34.613 34.5946 35.1268 35.8238 36.6519 37.968 39.1856
40.4048 42.0209 42.6411 43.839 45.3784 46.7911 48.1819 49.9876 51.7624
53.7624 55.7459 57.755 59.79 61.8312 63.764 65.8606 67.9603 70.0744 72.2157
74.3775 76.5159 78.6857 80.85 83.0304 85.2054 87.3873 89.5241 45.3055
46.0633 44.3405 44.3216 43.5937 43.3836 43.1729 43.0774 42.5224 41.9841
41.1101 40.2479 39.1479 37.9067 36.9188 35.5455 34.6196 33.0499 31.3119
30.4254 29.5232 27.9786 26.2643 25.8086 24.5383 23.5224 21.7715 20.2403
19.3241 18.5954 17.856 17.1438 16.2505 14.8663 13.8976 12.4935 10.747
9.07617 7.27777 5.58294 3.92122 2.17927 0.381098) 5->OrderedCollection (45
43.4307 41.7358 40.3057 39.3169 38.4369 37.6485 36.9528 36.5805 36.3011
35.9363 35.6996 35.6237 36.2089 36.8851 37.699 38.7534 39.083 40.35 41.2837
42.5761 44.2533 46.0166 47.7813 49.5788 51.4211 53.2502 55.0115 57.0217
59.0326 61.1127 63.1117 65.2053 66.8797 68.5816 70.2024 71.9561 71.2106
71.7644 71.1674 71.2402 72.4791 73.9797 74.9357 76.8403 77.6676 79.7265
81.8157 82.5328 82.9547 82.8807 82.4186 82.1191 82.2947 81.6566 81.3614
80.4294 79.7342 78.9221 77.5769 75.7841 74.6128 73.8142 73.154 72.2774
71.3685 70.5819 69.6826 68.8174 68.1853 66.5015 65.0468 64.0326 62.2445
61.3035 59.5129 58.5509 57.6963 56.5964 55.7203 54.2757 53.1037 51.9351
50.7397 49.9133 49.0981 48.3813 47.9851 46.9975 46.196 45.4027)
6->OrderedCollection (45 43.2262 41.4271 40.2103 39.1076 38.257 36.9944
36.3847 35.0641 34.5636 33.9294 33.0934 32.5598 32.2836 32.1609 32.3856
33.0671 33.4822 34.1303 34.9579 36.048 37.425 38.4235 39.5956 41.1531
42.8345 43.2804 43.9052 45.1672 47.1391 49.1093 51.1702 53.2206 55.1098
56.9685 58.8619 60.8772 61.8001 63.4326 65.5823 67.7313 69.8884 72.0283
74.2006 76.3451 78.4644 80.4807 82.4008 83.5171 84.9387 86.0579 86.9833
86.9778 86.5258 85.9836 84.9723 84.4694 83.6748 83.0443 82.4237 81.6418
81.0012 79.5479 78.0395 76.5218 75.6251 73.9764 72.1983 70.3985 68.6018
66.8024 65.1586 63.3618 61.6845 59.8923 58.1 56.3026 55.695 54.8657 53.4508
51.6513 49.8842 48.1689 46.3949 44.6616 43.8633 42.2974 40.7538 39.8139
38.6842 37.8159) 7->OrderedCollection (45 43.6572 41.9719 40.7301 39.4589
38.54 37.7499 37.0077 36.1815 35.6729 35.1885 34.9625 35.3337 35.7402
36.452 37.2713 38.0204 39.2948 40.1358 41.1642 42.274 43.5344 44.9117
46.4636 47.9712 49.4974 51.2098 53.0055 54.8536 56.8541 58.8918 60.9744
63.0558 65.1449 66.5137 67.8941 69.9558 71.8622 73.8008 75.8358 77.7945
79.8441 81.7555 82.4676 81.9716 81.9412 83.053 85.1307 85.6994 86.4264
87.1926 87.1548 86.0718 84.3781 82.9079 81.2383 79.6952 78.7672 77.7782
76.6391 75.4558 73.6779 72.381 70.6376 68.8434 67.0817 65.2871 63.5174
61.7181 60.0204 58.2212 56.4341 54.6343 52.8405 51.0689 49.269 47.4702
45.6773 43.9227 42.1424 40.3783 38.5784 36.7809 35.8783 34.0812 32.9605
32.0932 30.4552 28.6554 26.8592 25.8905) ...etc...) ) )

2018-06-01 14:18 GMT+02:00 Nicolas Becu <nicolas.becu@univ-lr.fr>:

> yes.
> global data that are stored with the model class key
> local probes data are stored  with the associated entity class key
>
> 2018-06-01 12:53 GMT+02:00 Serge Stinckwich <notifications@github.com>:
>
>> I try to understand the structure of the actual Dictionnary. You have
>> global data that are stored with the model class key and information about
>> each classes ? right ?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cormas/cormas/issues/56#issuecomment-393845758>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AHRtZyzKP-JfQg0gXeDxzkBAwUlx5Vptks5t4R0bgaJpZM4UVWUS>
>> .
>>
>
>
>
> --
> __________________
> Nicolas Becu
> Chercheur au CNRS
> Littoral Environnement et Sociétés (LIENSs) - UMR 7266
> Bâtiment ILE
> 2, rue Olympe de Gouges
> 17 000 La Rochelle
> Tel : 05 16 49 67 94
>



-- 
__________________
Nicolas Becu
Chercheur au CNRS
Littoral Environnement et Sociétés (LIENSs) - UMR 7266
Bâtiment ILE
2, rue Olympe de Gouges
17 000 La Rochelle
Tel : 05 16 49 67 94
